File: 프로그래머스/4/42891. 무지의 먹방 라이브/무지의 먹방 라이브.py
import heapq

def solution(food_times, k):
    if sum(food_times) <= k:
        return -1
    
    length = len(food_times)
    q = []
    for i in range(length):
        heapq.heappush(q, (food_times[i], i + 1))
        
    round = 0
    s = 0
    
    while s + ((q[0][0] - round) * length) <= k:
        now = heapq.heappop(q)[0]
        s += (now - round) * length
        length -=1
        round = now

    answer = sorted(q, key = lambda x: x[1])
    return answer[(k-s)%length][1]

File: 프로그래머스/4/42891. 무지의 먹방 라이브/무지의 먹방 라이브.java
import java.util.*;

class Solution {
    public int solution(int[] food_times, long k) {
        long sum = 0;
        
        for (int i = 0; i < food_times.length; i++) {
            sum += food_times[i];
        }
        
        if (sum <= k)
            return -1;
        
        PriorityQueue<Food> pq = new PriorityQueue();
        for (int i = 0; i < food_times.length; i++){
            pq.offer(new Food(food_times[i], i + 1));
        }
        
        sum = 0;
        long time = 0; // 먹은 음식 시간
        long length = food_times.length; // 회전판 길이
        
        while (sum + ((pq.peek().getTime() - time) * length) <= k) {
            int now = pq.poll().getTime();
            sum += (now - time) * length;
            length--;
            time = now;
        }
        
        ArrayList<Food> result = new ArrayList();
        
        while (!pq.isEmpty())
            result.add(pq.poll());
        
        Collections.sort(result, new Comparator<Food>(){
           public int compare(Food f1, Food f2){
               return Integer.compare(f1.getIndex(), f2.getIndex());
           } 
        });
        
        return result.get((int) ((k - sum) % length)).getIndex();
    }
}

class Food implements Comparable<Food>{
    int time;
    int index;
    
    public Food(int time, int index){
        this.time = time;
        this.index = index;
    }
    
    public int getTime(){
        return time;
    }
    
    public int getIndex(){
        return index;
    }
    
    public int compareTo(Food f){
        return Integer.compare(this.time, f.time);
    }
}